app/controllers/sessions_controller.rb

```diff
@@ -3,6 +3,31 @@ def new
     @user = User.new
   end
 
+  def login
+    client_id     =
+    client_secret = 
+    code          = params[:code]
+    response      = Faraday.post("https://github.com/login/oauth/access_token?client_id=#{client_id}&client_secret=#{client_secret}&code=#{code}")
+    pairs = response.body.split("&")
+    response_hash = {}
+    pairs.each do |pair|
+      key, value = pair.split("=")
+      response_hash[key] = value
+    end
+
+    token = response_hash["access_token"]
+    oauth_response = Faraday.get("https://api.github.com/user?access_token=#{token}")
+
+    auth = JSON.parse(oauth_response.body)
+    # user          = User.find_or_create_by(id: auth["id"])
+    user          = User.find(session[:user_id])
+    user.token    = "token #{token}"
+    # user.password = "password"
+    user.save!
+    session[:user_id] = user.id
+    redirect_to dashboard_path
+  end
+
```

Comment:
Followed the lesson plan word for word on Oauth. On line 24 I had to interpolate the token to have the correct format that GH needs.

Reply:
Also something of note is the commented-out line on 25. I was getting an error at one point that `password` cannot be blank. I commented-out the line later and everything was working but if you get a validation error that is probably the cause.